QSP transfer Lambda implementation

marcransome opened the pull request.

marcransome committed:
Add Lambda function implementation

marcransome committed:
Rename project

marcransome committed:
Add version file

marcransome committed:
Remove test suffix from remote filename

marcransome committed:
Extract file transfer logic to function

marcransome committed:
Whitespace cleanup

marcransome committed:
Improve exception handling

Jata2k commented on qsp_transfer.py:
```diff
@@ -0,0 +1,108 @@
+
+def get_ftp_credentials():
+    client = boto3.client('secretsmanager')
+    get_secret_value_response = client.get_secret_value(SecretId=os.environ['SECRET_NAME'])
```
We should probably do some exception handling here

marcransome replied:
I was originally catching and re-raising exceptions here in a `try/except` block but this was [simplified](https://github.com/companieshouse/ois-tuxedo-qsp-transfer/pull/1/commits/0c37f93242d0265028f039ca02eabef4c3a62da8) in favour of allowing exceptions to bubble up the call stack to be dealt with in the 'global' `try/except` block present in the handler function.

Since we don't need to respond to any particular exception with specific logic, all exceptions will be caught and logged in the handler and will include structured data in the log message for the error type, error message, and call stack. This also removes the need for additional `try/except` blocks here and in other places that may raise an exception.

We could reintroduce the `try/except` block here, and other places that may generate exceptions, but at the cost of some readability and some duplication. I'd be tempted to stick with the existing approach as the code is arguably cleaner and there's no functional need to respond to specific exceptions.

Jata2k commented on qsp_transfer.py:
```diff
@@ -0,0 +1,108 @@
+    if not re.match(r"\d{2}-\d{2}-\d{4}", date):
+        raise Exception(f"Date parameter value '{date}' does not match expected format DD-MM-YYYY")
+
+    log_group = os.environ['LOG_GROUP_NAME']
```
I'd use `getenv` here. In addition, I'd probably do some validation up front before we execute anything and potentially get into ambiguous states

marcransome replied:
I've added a `check_environment_variables` function to validate environment variables upfront using `os.getenv`.

I think it's worth maintaining the `os.environ` call here as this will raise a `KeyError` exception if execution ever got to this line and a mandatory variable is missing (e.g. if the variable was missed from the validation code in `check_environment_variables`) as `os.getenv` could return `None` here leading to some undefined behaviour.

Jata2k commented on qsp_transfer.py:
```diff
@@ -0,0 +1,108 @@
+        raise Exception(f"Date parameter value '{date}' does not match expected format DD-MM-YYYY")
+
+    log_group = os.environ['LOG_GROUP_NAME']
+    start_time = int(time.mktime(time.strptime(f"{date} 00:00:00", '%d-%m-%Y %H:%M:%S'))) * 1000
```
This feels like it could be pulled out into a separate function (`calculate_epoch_time`?) and refactored.

Jata2k commented on qsp_transfer.py:
```diff
@@ -0,0 +1,108 @@
+        endTime = end_time
+    )
+
+    events = events + response['events']
```
I'd use `append`. I.e. `events.append(response['events'])`

marcransome replied:
Agreed, this is currently allocating a new list every time rather than reusing the initial list, but `append` will result in nested lists so I'll replace this with `extend()`.

marcransome committed:
Minor refactoring improvements

- Add environment variable validation logic
- Extract epoch time handling logic into separate function
- Cater for scenarios where log events do not exist

marcransome committed:
Add type hints for documentation purposes

marcransome committed:
Simplify list length checks

marcransome committed:
Remove misleading type hint

marcransome committed:
Add type hint for return value

marcransome committed:
Add environment variable for remote path

Jata2k commented on qsp_transfer.py:
```diff
@@ -0,0 +1,136 @@
+            logger.info("No data for given time period; processing complete")
+        else:
+            transfer_data_file(path)
+            logging.info("Data file transfer completed")
```
It might be worth deleting the file once we're done with it as a precaution

marcransome committed:
Remove temp file to avoid stale content in subsequent invocations

marcransome committed:
Use consistent quotes
